compiler/src/dotty/tools/dotc/transform/init/ParamOverridingCheck.scala

```diff
@@ -0,0 +1,343 @@
+package dotty.tools.dotc
+package transform
+package init
+
+import core.*
+import Contexts.*
+import Symbols.*
+import Types.*
+import StdNames.*
+import NameKinds.OuterSelectName
+import util.Property
+
+import ast.tpd.*
+import config.Printers.init as printer
+import reporting.trace as log
+
+import Semantic.Arg
+import Semantic.NewExpr
+import Semantic.Call
+import Semantic.hasSource
+
+import scala.collection.mutable
+import scala.annotation.tailrec
+
+/**
+ * Check overriding of class parameters
+ *
+ * This check issues a warning if the use of a class pamameter in the primary
+ * constructor potentially has different semantics from its use in methods.
```

Comment:
I would suggest to include a more precise definition of the rule (probably found somewhere in the discussion of #15764). Just saying "different semantics" is vague. The rule should say what conditions a programmer has to satisfy in order to make the (conservative) warning go away. This is to help not only the programmer (and perhaps be put into the language spec at some point) but also to help people understand what exactly the code is checking for.

Reply:
Here's a specific comment that would help, taken from the discussion:

> The guarantee is the following:
> 
>   • A class parameter may only be overridden by another class parameter.
>   • If a parameter is overridden, at the end of class constructors, the
>     overridden and overriding parameters have the same value.